### DISCUSSION TITLE
ModuleNotFoundError

### DISCUSSION THREAD

**Discussion by ksabmagar7** [microsoft/PlanetaryComputer#205] [General]
Hello,

This must be beginner question.

How do we remove this frequent ModuleNotFoundError?
These are all happening in examples of planetary computer.
I have checked the resources on installing the the packages.
I can then install !pip install for the package name, but the error is the same.
I also open the terminal and repeat the process and import the package again, the error is same.
I have read that this is also due to the different versions of python.

How do well-versed people deal with this?
Any beginner friendly comments will be really help me in my learning/
I want to solve this for me.

Thankyou

**Comment by TomAugspurger** [comment]
In general, that error is raised when you (or some code you're using) tries to `import <some_module>` but `<some_module>` isn't installed.

Beyond that, it's hard to give too much guidance. Project Pythia has some good introductory material at https://foundations.projectpythia.org/foundations/getting-started-python.html. https://swcarpentry.github.io/python-novice-inflammation/index.html might be worth looking through too.

**Reply by John-Polo** [reply]
@TomAugspurger I have a similar problem with one of the data example notebooks. This one for Daymet: https://planetarycomputer.microsoft.com/dataset/daymet-annual-na#Example-Notebook

The first line of the first code chunk is

`import cartopy.crs as ccrs`

and that returns

> 
> ---------------------------------------------------------------------------
> ModuleNotFoundError                       Traceback (most recent call last)
> Cell In[1], line 1
> ----> 1 import cartopy.crs as ccrs
>       2 import fsspec
>       3 import matplotlib.pyplot as plt
> 
> ModuleNotFoundError: No module named 'cartopy'

I placed a new cell before that one and ran  

`pip install cartopy`

and that gives me this message

> /srv/conda/envs/notebook/bin/python: No module named pip
> Note: you may need to restart the kernel to use updated packages.

I comment out the cartopy line and then I get modulenotfounderrors for matplotlib and xarray. I suppose I don't need matplotlib, but xarray seems important for this task. So I don't know how to import necessary modules or install them. I'm not very experienced with Jupyter notebooks nor Python, but I don't think I've made any major errors here.

**Reply by TomAugspurger** [reply]
Hmm, just to double-check, you're running this from the Planetary Computer Hub? Which [environment](https://planetarycomputer.microsoft.com/docs/overview/environment/#logging-in) are you choosing when you log into the Hub?

**Reply by John-Polo** [reply]
My fault. I guess I had started with R. None of those errors occur. Thank you.

**Comment by ksabmagar7** [comment]
Hello,
Please suggest.
I have now learned to create the environment from the terminal, and also
activate that env. in the notebook.
But even after installing the package it still says the module is not
available.
I refresh it, rerun it. It does not work.
What is the mistake I am doing?
Thankyou
[image: image.png]
[image: image.png]

**Reply by TomAugspurger** [reply]
I'm not sure why. Maybe read through https://packaging.python.org/en/latest/tutorials/installing-packages/ for some background.
